### PR TITLE
 secboot,cmd/snap-bootstrap: cross-check partitions before unlocking, mounting

### PR DESCRIPTION
**Pull request by anonymouse64**
Based on top of #9170 for easier conflict resolution

Before we mount partitions, we should check / verify that the partition we want
to mount is coming from the same disk that we have already mounted partitions
from, or in the original case, that the partition comes from the same disk as we
booted the kernel from.

In the unlocking case in secboot, we should only try to unlock an encrypted
partition if it comes from the same disk that we mounted the first partition
from, i.e. ubuntu-boot in run mode and ubuntu-seed in recover mode.

Additionally, we should verify that the decrypted partition we got after
unlocking comes from the same encrypted partition that we requested it come
from.

Fixes: https://bugs.launchpad.net/snapd/+bug/1863886

### CODE REVIEW

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go [hunk 1]
```diff
@@ -413,7 +433,33 @@ func generateMountsModeRun(mst *initramfsMountsState) error {
...
 	}
 
-	// 4.1. read modeenv
+	// 4.1 verify that ubuntu-data comes from where we expect it to
+
+	// since ubuntu-data has already been mounted, see if there is an encrypted
+	// device that we should try to boot from
+	diskOpts := &disks.Options{}
+	_, err = disk.FindMatchingPartitionUUID("ubuntu-data-enc")
+	if err == nil {
```
Comment (pedronis) [inline cmd/snap-bootstrap/cmd_initramfs_mounts.go:442]: shouldn't we not ignore errors that are not finding the the partition with the name at all? see also later comment

File: secboot/secboot_tpm.go [hunk 2]
```diff
@@ -216,10 +217,22 @@ func UnlockVolumeIfEncrypted(name string, lockKeysOnFinish bool) (string, error)
...
-		ok, encdev := isDeviceEncrypted(name)
-		if !ok {
+		// find the encrypted device using the disk we were provided - note that
+		// we do not specify IsDecryptedDevice in opts because here we are
+		// looking for the encrypted device to unlock, later on in the boot
+		// process we will look for the decrypted device to ensure it matches
+		// what we expected
+		partUUID, err := disk.FindMatchingPartitionUUID(name + "-enc")
+		var errNotFound disks.FilesystemLabelNotFoundError
+		if xerrors.As(err, &errNotFound) {
```
Comment (pedronis) [inline secboot/secboot_tpm.go:227]: here we are doing this

File: secboot/secboot_tpm.go [hunk 3]
```diff
@@ -216,10 +217,22 @@ func UnlockVolumeIfEncrypted(name string, lockKeysOnFinish bool) (string, error)
...
+			// didn't find the encrypted label, so return nil to try the
+			// decrypted label again
 			return nil
```
Comment (pedronis) [inline secboot/secboot_tpm.go:230]: mmh, instead shouldn't we maybe simply return as well whether we decrpyted or not so that the caller doesn't need to check/guess again?

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go [hunk 4]
```diff
@@ -405,7 +434,33 @@ func generateMountsModeRun(mst *initramfsMountsState) error {
 		return err
 	}
 
-	// 4.1. read modeenv
+	// 4.1 verify that ubuntu-data comes from where we expect it to
+
+	// since ubuntu-data has already been mounted, see if there is an encrypted
+	// device that we should try to boot from
+	diskOpts := &disks.Options{}
+	_, err = disk.FindMatchingPartitionUUID("ubuntu-data-enc")
```
Comment (mvo5) [inline cmd/snap-bootstrap/cmd_initramfs_mounts.go:442]: (nitpick^2) This could be written in a slightly more compact way (single line): `if _, err := ...; err == nil {` which we tend to use a lot.

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go [hunk 5]
```diff
@@ -405,7 +434,33 @@ func generateMountsModeRun(mst *initramfsMountsState) error {
...
+
+	matches, err := disk.MountPointIsFromDisk(boot.InitramfsDataDir, diskOpts)
+	if err != nil {
+		return err
+	}
+	if !matches {
+		// failed to verify that ubuntu-data mountpoint comes from the same disk
+		// as ubuntu-boot
+		if diskOpts.IsDecryptedDevice {
+			return fmt.Errorf("cannot validate boot: ubuntu-data mountpoint is not from ubuntu-data-enc partition")
```
Comment (mvo5) [inline cmd/snap-bootstrap/cmd_initramfs_mounts.go:458]: I wonder if we should provide more error information in this case. Something like: `cannot validate boot: ubuntu-data mountpoint is from disk /dev/sda1 but from ubuntu-data-enc partition (/dev/sdb2)` or something like this? Only if we have the data of course. But giving some more hints is probably a good idea(?).

**Comment by anonymouse64** [comment]
Ah actually this is missing one bit, we should also verify that the decrypted /run/mnt/host/ubuntu-data comes from the expected encrypted partition too, since we copy important data from that mount point.

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go (hunk 5 above)
Comment (anonymouse64) [reply cmd/snap-bootstrap/cmd_initramfs_mounts.go:458]: I will reword this to be:

```
cannot validate boot: ubuntu-data mountpoint is expected to be from disk 252:1 but is not
```

because we don't currently have a way to get the /dev node for a disks.Disk, but we do have Dev() which returns the major/minor number. I like your suggestion though, and I think it would be easy to add this to the disks package but I'd rather do that in a followup to show /dev node instead of the device major/minor number. We also don't currently have a way to tell what disk a mount point is from at this point in the case they don't match, I think that probably we could modify the function signature to return what disk the mount point is from if it was found however that would also be a followup IMHO.

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go (hunk 4 above)
Comment (anonymouse64) [reply cmd/snap-bootstrap/cmd_initramfs_mounts.go:442]: sure, sorry I am more used to putting the err := ... on it's own line but I see that you prefer it always on the same line :smile:

Comment (anonymouse64) [reply cmd/snap-bootstrap/cmd_initramfs_mounts.go]: this bit actually ended up going away with Samuele's suggested refactor

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go (hunk 5 above)
Comment (anonymouse64) [reply cmd/snap-bootstrap/cmd_initramfs_mounts.go]: I added some TODOs and reworded the message, let me know if you want it reworded further

File: secboot/secboot_tpm.go (hunk 3 above)
Comment (anonymouse64) [reply secboot/secboot_tpm.go]: yes that ended up being fine to do

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go (hunk 1 above)
Comment (anonymouse64) [reply cmd/snap-bootstrap/cmd_initramfs_mounts.go]: with your other suggestion to return whether it was found or not from `secboot.Unlock...` this ended up being immaterial actually

**Comment by anonymouse64** [comment]
> Ah actually this is missing one bit, we should also verify that the decrypted /run/mnt/host/ubuntu-data comes from the expected encrypted partition too, since we copy important data from that mount point.

this is now done as well, it ended up not being too much more code on top of this, but I am re-requesting review just because it is an additional behavior change

**Comment by anonymouse64** [comment]
> The existing spread test covers that there are no regressions, yes?

Hmm to be sure there are not regressions, really this should also be run with the nested spread tests, so I will close and re-open to trigger those to run

**Comment by cmatsuoka** [comment]
Thanks Ian. It seems to be that we're now checking if the encrypted device resides in the right disk, and that the unencrypted partition actually comes from this encrypted partition if it exists, or from the right disk if encryption is not used. So I believe this will fix LP:1863886 and the TODO comment in `UnlockPartitionIfEncrypted` can be removed.

File: secboot/secboot_tpm_test.go [hunk 6]
```diff
@@ -377,20 +431,17 @@ func (s *secbootSuite) TestUnlockIfEncrypted(c *C) {
...
+		device, isDecryptDev, err := secboot.UnlockVolumeIfEncrypted(tc.disk, "name", boot.InitramfsEncryptionKeyDir, tc.lockRequest)
 		if tc.err == "" {
 			c.Assert(err, IsNil)
-		} else {
-			c.Assert(err, ErrorMatches, tc.err)
-		}
-		if tc.device == "" {
-			c.Assert(device, Equals, tc.device)
-		} else {
+			c.Assert(isDecryptDev, Equals, tc.hasEncdev)
```
Comment (cmatsuoka) [inline secboot/secboot_tpm_test.go:437]: I wondered if this was being tested but here it is :+1:

**Comment by anonymouse64** [comment]
The test failures here are unrelated to the PR, though some of them are puzzling. 

* `google-nested:ubuntu-20.04-64:tests/nested/manual/refresh-revert-fundamentals:snapd` is not understood why it failed, we couldn't login, but the VM serial log doesn't indicate that the VM went into a reboot loop or anything like that, so it's not clear what happened there, full log @ https://pastebin.ubuntu.com/p/Yskh9yqNZH/, I informed Sergio about this one
* `google-nested:ubuntu-20.04-64:tests/nested/manual/core20-early-config` is broken due to https://github.com/snapcore/snapd/pull/9102#discussion_r473185796, I will file a PR to fix that one
* `google-nested:ubuntu-16.04-64:tests/nested/core/core-snap-refresh-on-core` failed because the function `refresh_to_new_core` is currently written in a racy way, where it tries to talk to snapd after refreshing the core snap, but if snapd is quick enough it will shut itself down for the reboot and not produce any output to the `snap info` client
* `google-nested:ubuntu-16.04-64:tests/nested/core/extra-snaps-assertions` just timed out, happens sometimes with long lived spread nested runs
* `google:ubuntu-20.04-64:tests/main/lxd:*` are all the standard lxd failures right now
* `google:ubuntu-core-20-64:tests/smoke/` failed to come back after a reboot, we don't have more info/details unfortunately on this


As such, I would like to merge this PR, we have positive runs of other nested uc20 runs with encryption so this is good, as well as the normal uc20 main suite that passed many other tests. @mvo5 can you sudo merge this PR?
